### DISCUSSION TITLE
Two different types of neg_sampling for DIEN model.

### DISCUSSION THREAD

**Discussion by Jeriousman** [RUCAIBox/RecBole#1224] [Q&A] [answered]
As DIEN model combines CF and sequential, I know it should be a bit more complex in terms of correctly implementing it. 

To my knowledge, we need to do 2 types of negative samplings. 
First, of course we should do negative sampling for item sequence like we do in sequential models in Recbole. So we need to set correct feature that will be entailed with '_list' to define which feature is sequence type (I guess I am correct?).

Second, all the datapoints are considered to be positive examples in CTR prediction. So as long as I do not use any explicit label data, I will not have label feature column. In this case, I should mention in config 'neg_sampling' to do datapoint-wise for label negative sampling (Let me know if I am wrong). So I did all as I mentioned.

After data processing, I can see I got neg_item_id_list as expected, but I dont see label feature column in inter_feat or anywhere else. And the datapoint size is not augmented by 'neg_sampling' : {'uniform':2}.

So my questions are below.

1. Are my understandings are not correct?
2. How can I get CTR label as I expected and get datapoint-wise augmented dataset?

Thank you for your advice in advance.

**Comment by Ethan-TZ** [comment, marked as answer]
@Jeriousman Hello, thanks for your attention to RecBole!
You don't need to set correct feature that will be entailed with '_list' to define which feature is sequence type.

1. For DIEN, we implement speical interface to apply sequential data augmentation for both postive and negative sequential data. i.e., you just need to prepare the **.inter** in the form of `<user,item,timestamp>`，and we will generate the `item_id_list` and `neg_item_id_list`.
2. All the datapoints are considered to be positive examples. Therefore, you only need to keep the positive example in the dataset and filter out negative examples. E.g., for ml-1m, just set:
    ```
    val_interval:
      rating: "[3,inf)"
    ```

**Reply by Jeriousman** [reply]
Thank you for your feedback. I have a question about your second answer. As we all know, CTR datapoints are all positive examples. That is why we do negatve sampling for negative datapoints. In other libraries for DIEN, I checked they do negative sampling for datapoint-wise as well as negative sampling for sequential data. Why do you not do negative sampling datapoint-wise? Why do you only keep positive examples? How does it work? With only positive examples, we have just 'clicked' label in case of CTR. How do you solve this problem if we don't do negative sampling here?

Thank you.

**Reply by Ethan-TZ** [reply]
@Jeriousman Hello, DIEN and DIN are handled differently from other CTR models. For DIEN, we do negative sampling for datapoint-wise as well as negative sampling for sequential data. So the datapoints with 'non-clicked' are generated by negative sampling.

**Reply by Jeriousman** [reply]
Hello. Thank you for your feedback.
I guess I was not so clear. I said this

> After data processing, I can see I got neg_item_id_list as expected, but I dont see label feature column in inter_feat or anywhere else. And the datapoint size is not augmented by 'neg_sampling' : {'uniform':2}.

I can see neg sequences, but I dont see datapoint-wise negative sampling data. How can I check this? I do not see any label for 0/1 created, I dont see augmented datapoint-wise negatively sampled datapoints as well. Can you please guide me how to generate it and how to check the generated datapoint wise negative samples?

**Reply by Ethan-TZ** [reply]
@Jeriousman You can set breakpoint here to check the process of negative datapoints generation.
https://github.com/RUCAIBox/RecBole/blob/4f761699454fd5d8873607173d2514d38ac24031/recbole/data/dataloader/abstract_dataloader.py#L195

**Reply by Jeriousman** [reply]
I will look into it. Thank you.
